### PR TITLE
PLNSRVCE-1003: Pipeline service CI setup documentation

### PR DESCRIPTION
**Pull request by prietyc123**
This changes aims to add the documentation for pipeline service CI setup

Signed-off-by: Priti Kumari [pkumari@redhat.com](mailto:pkumari@redhat.com)

### CODE REVIEW

File: ci/docs/continuous_integation.md [hunk 1]
```diff
@@ -1,5 +1,20 @@
...
 
-The OpenShift robot is used to trigger functional tests on the code base.
+Pipelines as Code triggers the pipelineRun in [.tekton](https://github.com/openshift-pipelines/pipeline-service/tree/main/.tekton) which executes the functional tests tasks for each PR, before merging it.
+
+All the tests run on pipeline staging environment, where we spun up hypershift clusters.
+
+# Steps to setup Hypershift Configuration
+
+1. Follow [ROSA Provisioning](https://github.com/openshift-pipelines/pipeline-service/tree/main/ci/docs/rosa_cluster_provision.md) documentation for getting an AWS cluster.
+2. Intall and Configure [Hypershift](https://github.com/openshift-pipelines/pipeline-service/tree/main/ci/docs/hypershift_setup.md) on it.
```
Comment (Roming22) [inline ci/docs/continuous_integation.md:10]: `Install`

File: ci/docs/continuous_integation.md [hunk 2]
```diff
@@ -1,5 +1,20 @@
...
+
+# setup github app
+
+- Github app setup demo
```
Comment (Roming22) [inline ci/docs/continuous_integation.md:14]: Is this missing a link?

File: ci/docs/continuous_integation.md [hunk 3]
```diff
@@ -1,5 +1,20 @@
...
+
+# Bitwarden Secret
+
+Few secrets where we create HypershiftDeployment Resources has been stored in bitwarden like ROSA cluster kubeconfig, AWS pull secret and key pair.
```
Comment (Roming22) [inline ci/docs/continuous_integation.md:18]: `Bitwarden` (appears multiple times)

Comment (Roming22) [inline ci/docs/continuous_integation.md:18]: `have been stored`

Comment (Roming22) [inline ci/docs/continuous_integation.md:18]: `Few secrets where we create HypershiftDeployment Resources` -> not clear to me, can you rephrase?

File: ci/docs/continuous_integation.md [hunk 4]
```diff
@@ -1,5 +1,20 @@
...
+If you do not have bitwarden account, try creating one by following [Bitwarden Account Setup](https://redhat.service-now.com/help?id=kb_article_view&sys_kb_id=28b1450e975fd9102f8d361e6253af65)
```
Comment (Roming22) [inline ci/docs/continuous_integation.md:19]: Missing `.` at the end of the sentence.

File: ci/docs/hypershift_setup.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
+# Hypershift
+
+HyperShift is middleware for hosting OpenShift control planes at scale that solves for cost and time to provision, as well as portability across cloud service providers with strong separation of concerns between management and workloads. Clusters are fully compliant OpenShift Container Platform (OCP) clusters and are compatible with standard OCP and Kubernetes toolchains.
```
Comment (Roming22) [inline ci/docs/hypershift_setup.md:3]: Maybe make the first mention of Hypershift a link to the repo or project page?

File: ci/docs/hypershift_setup.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+# Steps to setup Hypershift on ROSA
+1. Login to the ROSA cluster created using [ROSA Provisioning](https://github.com/openshift-pipelines/pipeline-service/tree/main/ci/docs/rosa_cluster_provision.md) documentation.
+```bash
+oc login <API_URL>  -u <USERNAME> -p <PASSWORD>
+```
+2. Create MultiClusterEngine to enable Hypershift
+```yaml
+apiVersion: multicluster.openshift.io/v1
```
Comment (Roming22) [inline ci/docs/hypershift_setup.md:12]: Instead of documenting the manifests, maybe they should leave in the repository (maybe in `ci/manifests/hypershift`)?

File: ci/docs/hypershift_setup.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    cloud: auto-detect
+    vendor: auto-detect
+  name: local-cluster
+spec:
+  hubAcceptsClient: true
+  leaseDurationSeconds: 60
+```
+4. Create an S3 bucket for Hypershift Operator with public-read
+```bash
+$BUCKET_NAME=plnsvcci-oidc
```
Comment (Roming22) [inline ci/docs/hypershift_setup.md:38]: Instead of documenting the commands, maybe this can be put in a script (e.g. `ci/hack/hypershift_setup.sh`?)

The doc would become quite small and easy to follow, as it would explain what Hypershift, list the requirement(s) and then link to the script.

File: ci/docs/hypershift_setup.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+7. Create the secret in `ci-clusters` namespace
+```bash
+k create ns ci-clusters
+
+oc create secret generic my-aws-cred \
+  --from-literal=baseDomain='<base-domain>' \
+  --from-literal=aws_access_key_id='<aws key id>' \
+  --from-literal=aws_secret_access_key='<aws access key>' \
+  --from-literal=pullSecret='<Download your OpenShift Container Platform pull secret file from cloud.redhat.com/openshift/install/pull-secret by selecting Download pull secret.>' \
+  --from-file=ssh-publickey='/Users/username/.ssh/id_rsa.pub'
```
Comment (Roming22) [inline ci/docs/hypershift_setup.md:77]: `/User` is a macos path. It's better to use `$HOME`.

File: ci/docs/hypershift_setup.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+kubectl apply -f
+apiVersion: addon.open-cluster-management.io/v1alpha1
+kind: ManagedClusterAddOn
+metadata:
+  name: hypershift-addon
+  namespace: local-cluster
+spec:
+  installNamespace: open-cluster-management-agent-addon
+
+k -n local-cluster get ManagedClusterAddOn
```
Comment (Roming22) [inline ci/docs/hypershift_setup.md:62]: `k` is not a standard command.

File: ci/docs/rosa_cluster_provision.md [hunk 10]
```diff
@@ -0,0 +1,34 @@
+# Red Hat OpenShift Service on AWS (ROSA) Cluster Provisioning
+
+`rosa` is a command line tool that simplifies the use of Red Hat OpenShift Service on AWS, also known as ROSA. This cluster is required to manage Hypershift operator for pipeline-service.
```
Comment (Roming22) [inline ci/docs/rosa_cluster_provision.md:3]: The two sentences are disjointed. The second sentence starts with `This cluster`, the cluster has not been referenced before, only the ROSA service.

File: ci/docs/rosa_cluster_provision.md [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+Execute the following commands to provision a ROSA cluster:
+```bash
```
Comment (Roming22) [inline ci/docs/rosa_cluster_provision.md:6]: Same comment as the hypershift doc, you should move commands into scripts.

File: ci/docs/hypershift_setup.md (hunk 7 above)
Comment (prietyc123) [reply ci/docs/hypershift_setup.md:38]: Agree with you +1

**Comment by prietyc123** [comment]
I still need to add pac configuration steps for CI on this pr.

File: ci/hack/rosa_cluster_provision.sh [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+
+tmpdir=$(mktemp -d)
+cd "$tmpdir"
+
+# This repo has the rosa provision and destroy cluster scripts
+git clone git@github.com:stolostron/bootstrap-ks.git
+cd bootstrap-ks/rosa
+
+# Set the required environment variable before provisioning
+export ROSA_TOKEN=${ROSA_TOKEN:-"<your-rosa-token>"}
```
Comment (Roming22) [inline ci/hack/rosa_cluster_provision.sh:17]: Question: Do we want to read those values from a yaml and upload that yaml to Bitwarden? Or are those values something we do not even want to share at the team level?

**Comment by Roming22** [comment]
@prietyc123 Please resolve conversation when you submit an update that takes care of a comment.

File: ci/hack/rosa_cluster_provision.sh (hunk 12 above)
Comment (prietyc123) [reply ci/hack/rosa_cluster_provision.sh:17]: Yes we are not sharing those values at team level

File: ci/hack/hypershift_setup.sh [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    export BW_PASS=${BW_PASS:-"<your-bitwarden-master-password>"}
+    BW_SESSION=$(bw login "$BW_ACCOUNT" "$BW_PASS" --raw)
+    bw --session "$BW_SESSION" get password "hypershift-compute-kubeconfig" | base64 -d > hypershift-compute-kubeconfig
+    bw logout > /dev/null
+    unset BW_SESSION
+    unset BW_PASS
+    unset BW_ACCOUNT
+}
+
+echo "Get vault data from bitwarden"
```
[the remark this answers is not in the document]
Comment (prietyc123) [reply ci/hack/hypershift_setup.sh]: > Please add the -d and -h options of those scripts as well.

could you please elaborate `-d` and `-h`. Where you want me to add this.

**Comment by prietyc123** [comment]
> @prietyc123 Can you please address the comments so we can move forward?

@Roming22 Did I miss any comments which needs to be addressed? I think I addressed the comment you asked for. Could you please re-review.

File: ci/docs/continuous_integation.md [hunk 14]
```diff
@@ -1,5 +1,20 @@
 # Continuous integration tests
 
-The OpenShift robot is used to trigger functional tests on the code base.
+Pipelines as Code triggers the pipelineRun in [.tekton](https://github.com/openshift-pipelines/pipeline-service/tree/main/.tekton) which executes the functional tests tasks for each PR, before merging it.
```
[the remark this answers is not in the document]
Comment (Roming22) [reply ci/docs/continuous_integation.md]: `https://github.com/openshift-pipelines/pipeline-service` still appeasr multiple times. Please fix all occurences.

File: ci/docs/continuous_integation.md [hunk 15]
```diff
@@ -1,5 +1,28 @@
 # Continuous integration tests
 
-The OpenShift robot is used to trigger functional tests on the code base.
+Pipelines as Code triggers the PipelineRuns in [.tekton](../../.tekton) which executes the test tasks for each PR, before merging it.
```
Comment (Roming22) [inline ci/docs/continuous_integation.md:3]: `execute`

File: ci/docs/continuous_integation.md [hunk 16]
```diff
@@ -1,5 +1,28 @@
...
-TODO: add more information, link to the repository/folder controlling those tests.
+All the functional tests run on a hypershift aws cluster.
+
+# How to configure Hypershift
+
+Hypershift official [Documentation Guide](https://hypershift-docs.netlify.app/).
+
+## Pre-requisite:
+
+- install [bw](https://bitwarden.com/help/cli/) cli
```
Comment (Roming22) [inline ci/docs/continuous_integation.md:13]: `Install`

File: ci/docs/continuous_integation.md [hunk 17]
```diff
@@ -1,5 +1,28 @@
...
+- install [aws](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html) cli
```
Comment (Roming22) [inline ci/docs/continuous_integation.md:14]: `Install`.

File: ci/docs/continuous_integation.md (hunk 3 above)
Comment (Roming22) [reply ci/docs/continuous_integation.md]: Still needs to be fixed

File: ci/docs/rosa_cluster_provision.md [hunk 18]
```diff
@@ -0,0 +1,32 @@
+# Red Hat OpenShift Service on AWS (ROSA) Cluster Provisioning
+
+`rosa` is a command line tool that simplifies the use of Red Hat OpenShift Service on AWS, also known as ROSA.
+Visit official ROSA [documentation](https://access.redhat.com/products/red-hat-openshift-service-aws)
+
+We create a ROSA cluster on aws which manages Hypershift operator for pipeline-service.
```
[the remark this answers is not in the document]
Comment (Roming22) [reply ci/docs/rosa_cluster_provision.md]: `AWS` is still mispelled `aws` in various places.

File: ci/docs/rosa_cluster_provision.md [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+
+# How do I provision?
```
[the remark this answers is not in the document]
Comment (Roming22) [reply ci/docs/rosa_cluster_provision.md]: There are still issues with headers.

File: ci/hack/hypershift_setup.sh (hunk 13 above)
[the remark this answers is not in the document]
Comment (Roming22) [reply ci/hack/hypershift_setup.sh]: See the parse_args function in other scripts.

**Comment by Roming22** [comment]
@prietyc123 Please go through all open comments.

**Comment by prietyc123** [comment]
> There are still issues with headers.

I think you are looking into previous change only. Correct me if I am wrong. The current change has the headers you suggested.
image

Even the bitwarden stuffs you pointed out has been changed in the current change. Not sure whether the current change is visible to you or not 🤔

**Comment by prietyc123** [comment]
@Roming22 I have resolved all your concern. Could you please re-review. Thanks

File: ci/hack/rosa_cluster_provision.sh [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+init() {
+  tmpdir=$(mktemp -d)
+  cd "$tmpdir"  
+}
+
+provision_rosa_cluster() {
+  # This repo has the rosa provision and destroy cluster scripts
+  git clone git@github.com:stolostron/bootstrap-ks.git
```
Comment (bnallapeta) [inline ci/hack/rosa_cluster_provision.sh:85]: Always a best practice to checkout a branch or tag or commit to avoid surprises in the future in case they change something in main in the future.

File: ci/hack/hypershift_setup.sh [hunk 21]
```diff
@@ -0,0 +1,153 @@
...
+set -o errexit
+set -o nounset
+set -o pipefail
+
+usage() {
+    echo "
+Usage: 
+    ${0##*/} ./hypershift_setup.sh [options]
+
+Intsall Hypershift operator on ROSA cluster
```
Comment (bnallapeta) [inline ci/hack/hypershift_setup.sh:13]: Intsall -> Install

File: ci/docs/rosa_cluster_provision.md [hunk 22]
```diff
@@ -0,0 +1,32 @@
+# Red Hat OpenShift Service on AWS (ROSA) Cluster Provisioning
+
+`rosa` is a command line tool that simplifies the use of Red Hat OpenShift Service on AWS, also known as ROSA.
+Visit the official ROSA [documentation](https://access.redhat.com/products/red-hat-openshift-service-aws)
+
+We create a ROSA cluster on AWS which manages the Hypershift operator for pipeline-service.
+
+## How to provision a cluster?
+
+### Pre-requisite:
```
Comment (bnallapeta) [inline ci/docs/rosa_cluster_provision.md:10]: Pre-requisite: -> Pre-requisites:

File: ci/docs/continuous_integation.md [hunk 23]
```diff
@@ -1,5 +1,23 @@
...
+- Install [aws](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html) cli
+- Provision a ROSA cluster by following [documentation](https://github.com/openshift-pipelines/pipeline-service/tree/main/ci/docs/rosa_cluster_provision.md)
+
+### Hypershift Setup
+
+You can configure hypershift on ROSA by running the [hypershift_setup.sh](https://github.com/openshift-pipelines/pipeline-service/tree/main/ci/hack/hypershift_setup.sh) script.
+
+## Setup GitHub app
+
+- Need to configure GitHub app for Pipelines as Code configuration into pipeline-service repository.
```
Comment (bnallapeta) [inline ci/docs/continuous_integation.md:23]: pipeline-service -> Pipeline Service
(multiple occurrences)

File: ci/docs/continuous_integation.md [hunk 24]
```diff
@@ -1,5 +1,23 @@
...
+Pipelines as Code triggers the PipelineRuns in [.tekton](../../.tekton) which execute the test tasks for each PR, before merging it.
 
-TODO: add more information, link to the repository/folder controlling those tests.
+All the functional tests run on a hypershift AWS cluster.
+
+## How to configure Hypershift
+
+Hypershift official [Documentation Guide](https://hypershift-docs.netlify.app/).
+
+### Pre-requisite:
```
Comment (bnallapeta) [inline ci/docs/continuous_integation.md:11]: Pre-requisite: -> Pre-requisites:

File: ci/docs/continuous_integation.md [hunk 25]
```diff
@@ -1,5 +1,23 @@
 # Continuous integration tests
 
-The OpenShift robot is used to trigger functional tests on the code base.
+Pipelines as Code triggers the PipelineRuns in [.tekton](../../.tekton) which execute the test tasks for each PR, before merging it.
 
-TODO: add more information, link to the repository/folder controlling those tests.
+All the functional tests run on a hypershift AWS cluster.
+
+## How to configure Hypershift
```
Comment (bnallapeta) [inline ci/docs/continuous_integation.md:7]: Hypershift -> HyperShift
(multiple occurrences)

File: ci/docs/continuous_integation.md [hunk 26]
```diff
@@ -1,5 +1,23 @@
...
+
+Hypershift official [Documentation Guide](https://hypershift-docs.netlify.app/).
+
+### Pre-requisite:
+
+- Install [bw](https://bitwarden.com/help/cli/) cli
+- Install [aws](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html) cli
+- Provision a ROSA cluster by following [documentation](https://github.com/openshift-pipelines/pipeline-service/tree/main/ci/docs/rosa_cluster_provision.md)
```
Comment (Roming22) [inline ci/docs/continuous_integation.md:15]: Use local link.

File: ci/docs/continuous_integation.md [hunk 27]
```diff
@@ -1,5 +1,23 @@
...
+
+### Hypershift Setup
+
+You can configure hypershift on ROSA by running the [hypershift_setup.sh](https://github.com/openshift-pipelines/pipeline-service/tree/main/ci/hack/hypershift_setup.sh) script.
```
Comment (Roming22) [inline ci/docs/continuous_integation.md:19]: Use local link

File: ci/docs/rosa_cluster_provision.md [hunk 28]
```diff
@@ -0,0 +1,32 @@
...
+
+- AWS account secret access key and access id.
+- ROSA Token available [here](https://console.redhat.com/openshift/token/rosa)
+
+### Configure Cluster:
+
+You can configure a ROSA cluster using [rosa_cluster_provision.sh](https://github.com/openshift-pipelines/pipeline-service/tree/main/ci/hack/rosa_cluster_provision.sh) script.
```
Comment (Roming22) [inline ci/docs/rosa_cluster_provision.md:17]: Use local link

File: ci/docs/continuous_integation.md [hunk 29]
```diff
@@ -1,5 +1,23 @@
 # Continuous integration tests
 
-The OpenShift robot is used to trigger functional tests on the code base.
+Pipelines as Code triggers the PipelineRuns in [.tekton](../../.tekton) which execute the test tasks for each PR, before merging it.
 
-TODO: add more information, link to the repository/folder controlling those tests.
+All the functional tests run on a hypershift AWS cluster.
```
Comment (Roming22) [inline ci/docs/continuous_integation.md:5]: hypershift -> HyperShift (multiple occurences)

File: ci/hack/hypershift_setup.sh [hunk 30]
```diff
@@ -0,0 +1,153 @@
...
+      usage
+      exit 1
+      ;;
+    esac
+      shift
+  done
+}
+
+prechecks() {
+    KUBECONFIG="${KUBECONFIG:-}"
```
Comment (Roming22) [inline ci/hack/hypershift_setup.sh:91]: You are not checking for many of the mandatory arguments.

File: ci/hack/rosa_cluster_provision.sh [hunk 31]
```diff
@@ -0,0 +1,99 @@
...
+    esac
+      shift
+  done
+}
+
+init() {
+  tmpdir=$(mktemp -d)
+  cd "$tmpdir"  
+}
+
```
Comment (Roming22) [inline ci/hack/rosa_cluster_provision.sh:82]: You are not checking for any of the mandatory arguments.

Comment (prietyc123) [reply ci/hack/rosa_cluster_provision.sh:110]: Added

File: ci/hack/rosa_cluster_provision.sh (hunk 20 above)
Comment (prietyc123) [reply ci/hack/rosa_cluster_provision.sh]: @bnallapeta Could you please verify, if I understood you correctly.

Comment (bnallapeta) [reply ci/hack/rosa_cluster_provision.sh]: Looks good. :+1:

File: ci/docs/continuous_integation.md [hunk 32]
```diff
@@ -1,5 +1,28 @@
...
+
+You can configure HyperShift on ROSA by running the [hypershift_setup.sh](../hack/hypershift_setup.sh) script.
+
+## Setup GitHub app
+
+- Need to configure GitHub app for Pipelines as Code configuration into Pipeline Service repository.
+
+## Bitwarden secrets
+
+We import our hypershift secrets like pull-secret, public and private key pair, compute kubeconfig and base domain url to bitwarden.
```
Comment (Roming22) [inline ci/docs/continuous_integation.md:27]: HyperShift

File: ci/hack/hypershift_setup.sh [hunk 33]
```diff
@@ -0,0 +1,182 @@
...
+init() {
+  SCRIPT_DIR=$(
+    cd "$(dirname "$0")" >/dev/null
+    pwd
+  )
+}
+
+install_hypershift() {
+  echo "HyperShift setup on ROSA cluster"
+  # Enable HyperShift and make rosa cluster a managed cluster, visit documentation https://access.redhat.com/documentation/en-us/red_hat_advanced_cluster_management_for_kubernetes/2.6/html-single/multicluster_engine/index#hosted-control-planes-configure
```
Comment (Roming22) [inline ci/hack/hypershift_setup.sh:137]: ROSA

File: ci/hack/hypershift_setup.sh [hunk 34]
```diff
@@ -0,0 +1,182 @@
...
+  # Install HyperShift operator on the managed cluster
+  kubectl apply -f "$SCRIPT_DIR/ci/manifests/hypershift/hypershift_operator_install.yaml"
+  # Wait for HyperShift operator to be installed
+  while [ "$(kubectl -n local-cluster get ManagedClusterAddOn | grep -cE "hypershift-addon")" != "1" ]; do
+      echo -n "."
+      sleep 2
+  done
+  echo "HyperShift operator successfully installed on the managed cluster"
+
+  # Create aws credential secret
```
Comment (Roming22) [inline ci/hack/hypershift_setup.sh:161]: AWS

File: ci/hack/rosa_cluster_provision.sh [hunk 35]
```diff
@@ -0,0 +1,130 @@
...
+    fi
+}
+
+init() {
+  tmpdir=$(mktemp -d)
+  cd "$tmpdir"  
+}
+
+provision_rosa_cluster() {
+  # This repo has the rosa provision and destroy cluster scripts
```
Comment (Roming22) [inline ci/hack/rosa_cluster_provision.sh:112]: ROSA

**Review by Roming22 — COMMENTED** [review]
Still a few typos, but otherwise looks good.

**Comment by prietyc123** [comment]
> Still a few typos, but otherwise looks good.

Done.

**Review by Roming22 — APPROVED** [review]
